internal/tmpfs/archive.go

```diff
@@ -0,0 +1,253 @@
+	pd.Atime = uint32(d.openTime.Unix())
+	pd.Mtime = uint32(d.openTime.Unix())
+	pd.Length = 0
+	pd.Name = d.Name()
+	pd.User = uname
+	pd.Group = uname
+	return pd
+}
+
+func (a *Archive) Root() *Directory {
+	return a.root
+}
+
+func (a *Archive) addFile(filepath string, file *File) error {
+	filecmps := strings.Split(filepath, "/")
+
+	isFile := !file.hdr.FileInfo().IsDir()
+	var dirCmps []string
+	if isFile {
+		dirCmps = filecmps[:len(filecmps)-1]
+	} else {
+		dirCmps = filecmps
+	}
+
+	dir, err := a.getOrCreateDir(a.root, dirCmps)
+	if err != nil {
+		return err
+	}
+
+	if isFile {
+		return a.createFile(dir, filecmps[len(filecmps)-1], file)
+	}
+	return nil
+}
+
+func (a *Archive) getOrCreateDir(d *Directory, cmps []string) (*Directory, error) {
+	if len(cmps) == 0 {
+		return d, nil
+	}
+
+	cmpname := cmps[0]
+	if cmpname == "" {
+		return d, nil
+	}
+
+	if entry, exists := d.nameToEntryMap[cmpname]; exists {
+		if dir, ok := entry.(*Directory); ok {
+			return a.getOrCreateDir(dir, cmps[1:])
+		} else {
+			return nil, fmt.Errorf("File already exists with name %s", cmpname)
+		}
+	} else {
+		newDir := newDirectory(cmpname, d, a.openTime, uint64(len(a.dirs)))
+
+		a.dirs = append(a.dirs, newDir)
+
+		// Add the child dir to the parent
+		// Also serialize in p9 marshalled form so we don't need to faff around in Rread
+		d.nameToEntryMap[cmpname] = newDir
+		d.entries = append(d.entries, newDir)
+
+		return a.getOrCreateDir(newDir, cmps[1:])
+	}
+}
+
+func (a *Archive) createFile(d *Directory, filename string, file *File) error {
+	if _, exists := d.nameToEntryMap[filename]; exists {
+		return fmt.Errorf("File or directory already exists with name %s", filename)
+	}
+	d.nameToEntryMap[filename] = file
+	d.entries = append(d.entries, file)
+
+	file.fileQid.Type = protocol.QTFILE
+	file.fileQid.Path = uint64(len(a.files) + 1)
+
+	a.files = append(a.files, file)
+
+	return nil
+}
+
+func (a *Archive) DumpEntry(e Entry, parentPath string) {
+	if dir, isDir := e.(*Directory); isDir {
+		parentPath = path.Join(parentPath, dir.name)
+		for _, child := range dir.entries {
+			a.DumpEntry(child, parentPath)
+		}
+	} else if file, isFile := e.(*File); isFile {
+		log.Printf("%v\n", path.Join(parentPath, file.name))
+	}
+}
+
+func (a *Archive) DumpArchive() {
+	a.DumpEntry(a.root, "")
+}
+
+// ReadImage reads a compressed tar to produce a file hierarchy
+func ReadImage(r io.Reader) *Archive {
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := gzr.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	openTime := time.Now()
+	fs := &Archive{newDirectory("/", nil, openTime, 0), []*Directory{}, []*File{}, openTime}
+	tr := tar.NewReader(gzr)
+	for id := 0; ; id++ {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		file := newFile(hdr, uint64(id), uint64(hdr.Size))
+		if _, err := io.ReadFull(tr, file.data); err != nil {
```

Comment:
so do we want to have the entire contents of the archive, uncompressed, in-memory?

Reply:
at this point, I don't see the harm.  If it gets too large, or we do care, it can be changed.